mmcv/runner/hooks/logger/pavi.py

```diff
@@ -119,25 +147,26 @@ def log(self, runner) -> None:
 
     @master_only
     def after_run(self, runner) -> None:
+
         if self.add_last_ckpt:
-            ckpt_path = osp.join(runner.work_dir, 'latest.pth')
-            if osp.islink(ckpt_path):
-                ckpt_path = osp.join(runner.work_dir, os.readlink(ckpt_path))
+            # using runner.epoch/iter is ok since the step has been + 1
+            step = runner.epoch if self.by_epoch else runner.iter
 
-            if osp.isfile(ckpt_path):
-                # runner.epoch += 1 has been done before `after_run`.
-                iteration = runner.epoch if self.by_epoch else runner.iter
-                return self.writer.add_snapshot_file(
-                    tag=self.run_name,
-                    snapshot_file_path=ckpt_path,
-                    iteration=iteration)
+            ckpt_path = osp.join(runner.work_dir, 'latest.pth')
+            self._add_ckpt(runner, ckpt_path, step)
 
         # flush the buffer and send a task ending signal to Pavi
         self.writer.close()
 
     @master_only
     def before_epoch(self, runner) -> None:
-        if runner.epoch == 0 and self.add_graph:
+        super().before_epoch(runner)
+
+        step = self.get_epoch(runner) if self.by_epoch else self.get_iter(
+            runner)
```

Comment:
```suggestion
        step = self.get_epoch(runner)
```

Reply:
The action to add graph should be replaced in `before_train_epoch` or `before_train_iter`.